src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,226 @@
+section \<open>Chain-Sync Mini-Protocol\<close>
+
+text \<open>
+  The theory in this section implements the chain synchronization mini-protocol.
+\<close>
+
+theory "Ouroboros-Mini_Protocols-Chain_Sync"
+  imports
+    "Ouroboros-Mini_Protocols"
+    "HOL-Library.BNF_Corec"
+begin
+
+subsection \<open>Cryptographic Primitives\<close>
+
+typedecl hash
+
+axiomatization
+where
+  hash_linear_order: "OFCLASS(hash, linorder_class)"
+
+instance hash :: linorder
+  by (fact hash_linear_order)
+
+subsection \<open>Blocks\<close>
+
+datatype block = Block (hash: \<open>hash\<close>)
```

Comment:
I understand that defining blocks as being equivalent to hashes is preliminary, possibly because we lack some precise idea what blocks actually contain. That said, I would approach this area differently.

First, the contents of a block determine its hash. This should be reflected in the specification, I guess, in that hashes aren’t modeled as parts of blocks but rather as something derived from blocks via some function.

Second, for the sake of chain synchronization, it’s even irrelevant what a block contains. What’s more, chain synchronization is supposed to be polymorphic in the transferred data items, making it possible to not only transfer full blocks but also just headers.

That said, block headers differ from full blocks in that they do *not* determine block hashes. Or am I mistaken here?

All in all, I suggest that we introduce a type parameter for data items to be transferred (blocks, headers, or whatever) and maintain hashes separately from such data items and not as parts of them.

Reply:
As you correctly stated, the chain synchronization mini-protocol should be agnostic in the data transferred, which is supposed to be full blocks in the node-to-client scenario and headers in the node-to-node scenario. Moreover, as I understand from the specification document, the use of hashes, similar to the case of slots, is just an implementation detail (e.g., hashes are normally part of  headers and of points). In the Appendix A of the specification document, the types `block`, `header`, `tip` and `point` (the ones used by the mini-protocol) are defined as `any`, i.e., as unspecified types. Therefore, I think we should also get rid of hashes and instead use only the concepts "item", "tip", "point" and lists thereof. Of course, this may imply that certain functions which have implementations in our current code must be turned into abstract functions, i.e., additional parameters of the `chain_sync` locale, possibly including assumptions on their expected behaviour. What are your thoughts on this?

Reply:
I think I need some clarification, since I’m not so deeply into all these things.

First about hashes. My understanding is that a hash is computed from a block, which consist of a header and a body. Is the hash also stored as part of the header? If yes, doesn’t this imply that for computing a hash you already need to know the hash you want to compute, since it is part of what you want to hash? If a hash is part of the header (and thus, I guess, also part of the whole block), then we could and probably should not store the hash along the data item but rather specify that a hash can be computed from a data item.

Then there are points. What are these actually? Are they hashes? If yes, then we shouldn’t use different types for hashes and points, I’d say, because this makes the formalization quite complicated.

Actually, your formalizations generally tend to have a lot of types and constants, which might make them a bit hard to understand, as the reader has to familiarize himself with all of these. Abstraction is good, of course, but if a point is really guaranteed to be a hash, then, I guess, it’s better to use the type of hashes directly and perhaps use `point` as the name of a function that yields a point.

If we unify types this way, the question is which name we should prefer for our single type: `hash` or `point`. Actually, I don’t like both very much. “Hash” is perhaps a bit too concrete, since in our chain synchronization specification we don’t actually model computation of hashes, at least not directly. “Point” refers to a particular use of hashes: referencing points in a chain. Something like “ID” might be ideal (pun semi-indented :slightly_smiling_face:), but it’s probably not good to add yet another term. Maybe we should go with “hash”, which is at least something the other guys are very much familiar with. After all, I suppose that IDs of blocks are necessarily hashes of some form: they probably have to be determined entirely by the content they refer to.

Finally, what does the “tip” type stand for? Is this again just an alias for “hash”? What is it used for?

Reply:
Hmm, do I remember this correctly that @dcoutts once mentioned that points are hashes paired with something else, an index into the chain perhaps? If that’s the case, then we should probably name our single type `point` and not specify that it can be derived from a data item but rather store it along data items. But we should discuss this first.

Reply:
Or maybe we *should* use something like `block_id`, which would mean using a new term, of course, but at least make it somehow clear what we really mean (“point” is really a bit unspecific, in my opinion).

Reply:
From what I can see, the section on the chain synchronization protocol in *The Shelley Networking Protocol* doesn’t talk about hashes except for presenting some hashes in the examples. On the other hand, it *does* talk about points and tips. I guess, given this situation, it might be best to have a type parameter `'point` and no type parameter `'hash`. Since it *might* be the case that a point cannot be derived from the block it’s referring to, we might want to store points alongside data items instead of claiming that the former can be derived from the latter. Even if they can be derived, not stating this possibility might be good, as it make the specification more abstract. What do you think?

Reply:
First of all, I'd like to point out that my "extra knowledge" about some things comes from my work on other formalizations (e.g., the Ouroboros Praos specification) and from sporadic inspections of parts of the Haskell implementation. However, I think we should try to stick to what the specification document says (although, as we already know, the document is a bit vague in several parts)

Regarding hashes, the specification document doesn't mention hashes at all in the section on the chain synchronization mini-protocol, except for the depiction of headers/blocks in the diagrams. That's why I wonder whether hashes are actually needed in our specification or we can stay at a higher-level of abstraction and leave them out. Now, as for your questions: It's my understanding that a block _header_ includes two hashes: The hash of the previous block's header and the hash of the block body (which basically contains a list of transactions). Therefore, a block does not contain its own hash.

Now, regarding points: The specification document does not define what a point is, although I'm pretty sure our current understanding is correct, i.e., that a point refers to a specific block in a specific chain. By looking at the implementation, a point seems to be defined as a pair of a hash (the block's header hash, I presume) and a slot number.

Finally, regarding tips: Again, the specification document does not define what a tip is, but it's pretty clear to me that it refers to the most recent block/header in a chain (although I don't know if it's actually a block/header, a point to it, or something else). Anyhow, tips don't seem to be used, although they are included in most of the messages.

Reply:
> What do you think?
Absolutely agreed. That's perfectly in line what I have mentioned in my comment above and what I have already starting to implement. :slightly_smiling_face:

Reply:
Also, regarding tips, I think we should either

1. Add a new type parameter `'tip` (since we don't what the structure of a tip is) and a new function parameter to compute a tip from a collection of items; or
2. Drop tips altogether.

What do you think?

Reply:
Thank you very much for your elaborate response. I think I have a better understanding of the situation now and we’re pretty much on the same page already. Let me state my recommendations regarding how to proceed:

* We do not talk about hashes at all.
* We use the notion of point and leave it completely abstract, meaning we have a type parameter `'point` and no assumptions regarding it.
* At least for now, we do not deal with tips at all.

How does this sound?

Reply:
That sounds very good to me, let's go for it.